src/libostree/ostree-bootloader-grub2.c

```diff
@@ -315,7 +316,45 @@ _ostree_bootloader_grub2_write_config (OstreeBootloader      *bootloader,
         use_system_grub2_mkconfig = FALSE;
     }
   else
-    grub_exec = use_system_grub2_mkconfig ? GRUB2_MKCONFIG_PATH : "/usr/lib/ostree/ostree-grub-generator";
+    {
+      if (use_system_grub2_mkconfig)
+        {
+          grub_exec = GRUB2_MKCONFIG_PATH;
+        }
+      else
+        {
+          g_autoptr(GPtrArray) boot_loader_configs = NULL;
+          OstreeBootconfigParser *config;
+          g_autofree char *ostree_arg = NULL;
+          g_autofree char *sysroot_abspath = g_file_get_path (self->sysroot->path);
+          struct stat stbuf;
+
+          if (!_ostree_sysroot_read_boot_loader_configs (self->sysroot, bootversion, &boot_loader_configs,
+                                                         cancellable, error))
+            goto out;
+
+          config = boot_loader_configs->pdata[0];
+          ostree_arg = get_ostree_kernel_arg_from_config (config);
+          if (!ostree_arg)
+            {
+              g_set_error_literal (error, G_IO_ERROR, G_IO_ERROR_FAILED,
+                                   "No ostree= kernel argument found in boot loader configuration file");
+              goto out;
+            }
+
+          snprintf (generator_path, sizeof(generator_path), "%s/%s/usr/lib/ostree/ostree-grub-generator",
```

Comment:
The problem with this in the general case is that it assumes the new tree's binaries work on the current one.  For say ELF binaries, if the new binary has a new dependency (e.g. versioned glibc symbol), or a new shared library, it's going to fail.  In this case it's a shell script so that'll work fine, but *everything else* it executes is going to be from the current tree...

There are going to be cases where we need to run code from the new tree - I have pending work on recompiling SELinux policy, and was going to look at doing that via a container (`bwrap`/`systemd-nspawn`) in the new root.

But what you're doing here isn't strictly necessary, right?  It's just trying to avoid the user having to upgrade twice to get the change, but at a risk of more incompatibilities.